### PR TITLE
CDAP-7910 Hide entities in metadata search which starts with underscore

### PR DESCRIPTION
**Pull request by rohitsinha54**
Issue: https://issues.cask.co/browse/CDAP-7910
Build: http://builds.cask.co/browse/CDAP-RUT405-1
ITM: http://builds.cask.co/browse/CDAP-ITM6-12

Note: This will need doc change in the Metadata Search API to tell user that they can see entities whose name starts with _ by specifying showHidden=true. Doc change will come in another PR.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java [hunk 1]
```diff
@@ -170,9 +168,11 @@ void addTags(NamespacedEntityId namespacedEntityId, String... tags)
    * @param cursor the cursor that acts as the starting index for the requested page. This is only applicable when
    *               #sortInfo is not {@link SortInfo#DEFAULT}. If offset is also specified, it is applied starting at
    *               the cursor. If {@code null}, the first row is used as the cursor
+   * @param showHidden boolean which specifies whether to display hidden entities (entity whose name start with "_")
+   *                    or not. Default to false.
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java]: how is there a default. Looks like this is a required parameter

File: cdap-client-tests/src/test/java/co/cask/cdap/client/MetadataHttpHandlerTestRun.java [hunk 2]
```diff
@@ -1256,41 +1257,59 @@ public void testSearchResultPaginationWithTargetType() throws Exception {
...
+      mydataset,
       new DatasetInstanceConfiguration(Table.class.getName(), Collections.<String, String>emptyMap())
     );
 
     // create entities so system metadata is annotated
     streamClient.create(stream1);
     streamClient.create(stream2);
 
-    String sort = AbstractSystemMetadataWriter.ENTITY_NAME_KEY + " asc";
+    // do sorting with creation time here since the testSearchResultPagination does with entity name
```
Comment (anew) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/MetadataHttpHandlerTestRun.java:1280]: is changing of the sort order a fix unrelated to the title of this PR?

File: cdap-common/src/main/resources/cdap-default.xml [hunk 3]
```diff
@@ -403,7 +403,7 @@
 
   <property>
     <name>audit.enabled</name>
-    <value>true</value>
+    <value>false</value>
```
Comment (anew) [inline cdap-common/src/main/resources/cdap-default.xml]: unrelated change? Why the change?

**Review by anew — COMMENTED** [review]
changes related to skipping hidden entities look good. I am not sure, but there appear to be unrelated changes, too. Which I don't know whether they are intended...

File: cdap-client-tests/src/test/java/co/cask/cdap/client/MetadataHttpHandlerTestRun.java (hunk 2 above)
Comment (rohitsinha54) [reply cdap-client-tests/src/test/java/co/cask/cdap/client/MetadataHttpHandlerTestRun.java:1280]: In the earlier PR where this was added there was some discrepancy in the intention and code https://github.com/caskdata/cdap/pull/7427/files#diff-b4114e6a647587816791abcd9ef15f72R1263
The comment here was itending to do test with creation time but the sorting was based on entity-name. We already have another test which was testing with entity-name so here we are doing sorting with creation-time to cover both sorting possibility.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java (hunk 1 above)
Comment (rohitsinha54) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataAdmin.java]: Fixed. Also removed from other places where this is not true.

File: cdap-client-tests/src/test/java/co/cask/cdap/client/MetadataHttpHandlerTestRun.java (hunk 2 above)
Comment (anew) [reply cdap-client-tests/src/test/java/co/cask/cdap/client/MetadataHttpHandlerTestRun.java:1280]: oh ok

**Comment by anew** [comment]
LGTM

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java [hunk 4]
```diff
@@ -840,7 +840,8 @@ public void searchMetadata(HttpRequest request, HttpResponder responder,
                              // 2147483647 is Integer.MAX_VALUE
                              @QueryParam("limit") @DefaultValue("2147483647") int limit,
                              @QueryParam("numCursors") @DefaultValue("0") int numCursors,
-                             @QueryParam("cursor") @DefaultValue("") String cursor) throws Exception {
+                             @QueryParam("cursor") @DefaultValue("") String cursor,
+                             @QueryParam("showHidden") boolean showHidden) throws Exception {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: don't need a ``@DefaultValue``? does it default to false if the param is not specified?

File: cdap-client-tests/src/test/java/co/cask/cdap/client/MetadataHttpHandlerTestRun.java [hunk 5]
```diff
@@ -1256,41 +1257,59 @@ public void testSearchResultPaginationWithTargetType() throws Exception {
...
+    // though it was created before since showHidden is false and it should not affect pagination
+    searchResponse = searchMetadata(namespace, "*", ImmutableSet.of(MetadataSearchTargetType.DATASET),
+                                    sort, 0, 2, 2, null);
+    expectedResults = ImmutableList.of(new MetadataSearchResultRecord(mydataset));
+    expectedCursors = ImmutableList.of();
+    Assert.assertEquals(expectedResults, new ArrayList<>(searchResponse.getResults()));
+    Assert.assertEquals(expectedCursors, searchResponse.getCursors());
   }
 
+  @Test
```
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/MetadataHttpHandlerTestRun.java:1304]: ummm - this annotation was missing?

**Review by bdmogal — COMMENTED** [review]
@rohitsinha54 ``MetadataSearchResponse`` should contain all the parameters in the query. Should we add ``showHidden`` there too?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java (hunk 4 above)
Comment (rohitsinha54) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: Yes.

File: cdap-client-tests/src/test/java/co/cask/cdap/client/MetadataHttpHandlerTestRun.java (hunk 5 above)
Comment (rohitsinha54) [reply cdap-client-tests/src/test/java/co/cask/cdap/client/MetadataHttpHandlerTestRun.java:1304]: It was there initially. It got removed accidentally in some other PR.

**Comment by rohitsinha54** [comment]
```
@rohitsinha54 MetadataSearchResponse should contain all the parameters in the query. Should we add showHidden there too?
```
I am not sure about this. Let's take @anew opinion and I will add it if needed.

**Comment by rohitsinha54** [comment]
Since, we are on crunch time I added the change to include showHidden in the search response so that we can have build etc passed if we want to have it. Although, I will wait on merge for @anew confirmation as whether we should show it in response or not.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java (hunk 4 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: what's the defaultValue? always required? o.0

Comment (anwar6953) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: seems bad to always require it.

Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: I think it was accidentally removed in response to another comment from me.

**Comment by anew** [comment]
I don't see a harm in including the showHidden in the response. It is not clear to me how this breaks the build, though. I am ok with including it. Please add back the defaultValue in the @QueryParam and then LGTM

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java (hunk 4 above)
Comment (rohitsinha54) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: Since its a boolean it defaults to false. Please see conversation here: https://github.com/caskdata/cdap/pull/7439#discussion_r92915860
It's not always required. If not specified it will be false.
It seems like it's causing confusion about its default value. So, maybe explicitly saying it will be better. Although, it will be redundant. @anew Should we explicitly say it ?

**Comment by rohitsinha54** [comment]
Build passed here: 
http://builds.cask.co/browse/CDAP-RUT405-3
ITM passed here: 
http://builds.cask.co/browse/CDAP-ITM6-12

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java (hunk 4 above)
Comment (anwar6953) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: It wasn't clear to me that this would default to false (since its a query param, and didn't have the DefaultValue annotation). I think its better to be explicit in this case.
See the parameter numCursors just 2 lines above.
It's explicit, even though the default value is 0. This inconsistency can be confusing.

Comment (anwar6953) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java]: https://github.com/caskdata/cdap/pull/7439#discussion_r92953738
